src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/DesignTimeInputsFileWatcher.cs

```diff
@@ -0,0 +1,185 @@
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Threading;
+using System.Threading.Tasks.Dataflow;
+
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Threading;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.TempPE
+{
+    /// <summary>
+    /// Produces output whenever a design time input changes
+    /// </summary>
+    [Export(typeof(IDesignTimeInputsFileWatcher))]
+    [AppliesTo(ProjectCapability.CSharpOrVisualBasicLanguageService)]
+    internal class DesignTimeInputsFileWatcher : ProjectValueDataSourceBase<string[]>, IVsFreeThreadedFileChangeEvents2, IDesignTimeInputsFileWatcher
+    {
+        private readonly IProjectThreadingService _threadingService;
+        private readonly IDesignTimeInputsDataSource _designTimeInputsDataSource;
+        private readonly IVsService<IVsAsyncFileChangeEx> _fileChangeService;
+        private ImmutableDictionary<string, uint> _fileWatcherCookies = ImmutableDictionary<string, uint>.Empty.WithComparers(StringComparers.Paths);
+        private readonly SemaphoreSlim _disposeGate = new SemaphoreSlim(initialCount: 1);
+
+        private int _version;
+        private IDisposable? _dataSourceLink;
+
+        /// <summary>
+        /// The block that receives updates from the active tree provider.
+        /// </summary>
+        private IBroadcastBlock<IProjectVersionedValue<string[]>>? _broadcastBlock;
+
+        /// <summary>
+        /// The public facade for the broadcast block.
+        /// </summary>
+        private IReceivableSourceBlock<IProjectVersionedValue<string[]>>? _publicBlock;
+
+        [ImportingConstructor]
+        public DesignTimeInputsFileWatcher(ConfiguredProject project,
+                                           IDesignTimeInputsDataSource designTimeInputsDataSource,
+                                           IVsService<SVsFileChangeEx, IVsAsyncFileChangeEx> fileChangeService)
+             : base(project.Services, synchronousDisposal: true, registerDataSource: false)
+        {
+            _threadingService = project.Services.ThreadingPolicy;
+            _designTimeInputsDataSource = designTimeInputsDataSource;
+            _fileChangeService = fileChangeService;
+        }
+
+        public override NamedIdentity DataSourceKey { get; } = new NamedIdentity(nameof(DesignTimeInputsFileWatcher));
+
+        public override IComparable DataSourceVersion => _version;
+
+        public override IReceivableSourceBlock<IProjectVersionedValue<string[]>> SourceBlock
+        {
+            get
+            {
+                EnsureInitialized();
+
+                return _publicBlock!;
+            }
+        }
+
+        protected override void Initialize()
+        {
+            base.Initialize();
+
+            _dataSourceLink = _designTimeInputsDataSource.SourceBlock.LinkToAsyncAction(ProcessDesignTimeInputs);
```

Comment:
This has a race, your callback can be called before the public/broadcast block exist.